File: project_euler/java/problem6.java
/*
Sum square difference
Problem 6

The sum of the squares of the first ten natural numbers is,
12 + 22 + ... + 102 = 385

The square of the sum of the first ten natural numbers is,
(1 + 2 + ... + 10)2 = 552 = 3025

Hence the difference between the sum of the squares 
of the first ten natural numbers and the square of 
the sum is 3025 − 385 = 2640.

Find the difference between the sum of the squares of 
the first ONE HUNDRED natural numbers and the square of the sum.

SOLUTION: 2.516415E7 = 25164150
*/

/* 
Closed Form for square of sum
1 + 2 + ... + N = N*(N+1)/2, so
(1 + ... + N)**2 = (N*(N+1))**2 / 4

Closed Form for sum of squares
1**2 + ... + N**2 = N*(N+1)*(2*N+1)/6
*/

public class problem6 {
	public static void main(String[] args) {
		int N = 100;
		double sum_of_squares = N*(N+1)*(2*N+1)/6;
		double square_of_sum = Math.pow(N*(N+1), 2) / 4;
		System.out.println(square_of_sum - sum_of_squares);
	}
}

File: project_euler/python/problem6.py
"""
Sum square difference
Problem 6

The sum of the squares of the first ten natural numbers is,
1^2 + 2^2 + ... + 10^2 = 385

The square of the sum of the first ten natural numbers is,
(1 + 2 + ... + 10)^2 = 55^52 = 3025

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.
-------------------------------------------------------
STATUS: COMPLETED
ANSWER: 25164150
RUNTIME: 0.4s
COMPLEXITY: O(n)
-------------------------------------------------------
"""

def sum_squares(a,b):
    """
    Returns the sum of the square of all 
    numbers between a and b inclusive
    """
    return sum([x**2 for x in range(a,b+1)])

def square_sum(a,b):
    """
    Returns the the square of the
    sum of all the numbers from a 
    to b inclusive
    """
    return (sum([x for x in range(a,b+1)])) ** 2


start = 1
end = 100
print(square_sum(start, end) - sum_squares(start, end))